src/studio/src/designer/backend/Services/Implementation/SchemaModelService.cs

```diff
@@ -68,26 +89,106 @@ public async Task UpdateSchema(string org, string repository, string developer,
         {
             var altinnGitRepository = _altinnGitRepositoryFactory.GetAltinnGitRepository(org, repository, developer);
 
-            if (altinnGitRepository.GetRepositoryType().Result == Enums.AltinnRepositoryType.App)
+            if (await altinnGitRepository.GetRepositoryType() == AltinnRepositoryType.App)
             {
                 await UpdateAllAppModelFiles(org, repository, developer, relativeFilePath, jsonContent);
             }
             else 
             {              
                 await altinnGitRepository.WriteTextByRelativePathAsync(relativeFilePath, jsonContent, true);
-            } 
+            }
+        }
+
+        /// <summary>
+        /// Builds a JSON schema based on the uploaded XSD.
+        /// </summary>
+        /// <remarks>
+        /// This operation is using the new data modelling library.
+        /// </remarks>
+        /// <param name="org">Organization owning the repository identified by it's short name.</param>
+        /// <param name="repository">Repository name to search for schema files.</param>
+        /// <param name="developer">Developers short name</param>
+        /// <param name="fileName">The name of the file being uploaded.</param>
+        /// <param name="xsdStream">Stream representing the XSD.</param>
+        public async Task<string> BuildSchemaFromXsd(
+            string org, string repository, string developer, string fileName, Stream xsdStream)
+        {
+            var altinnAppGitRepository = _altinnGitRepositoryFactory.GetAltinnAppGitRepository(
+                org, repository, developer);
+
+            MemoryStream xsdMemoryStream = new MemoryStream();
+            xsdStream.CopyTo(xsdMemoryStream);
+
+            AltinnRepositoryType altinnRepositoryType = await altinnAppGitRepository.GetRepositoryType();
+
+            if (altinnRepositoryType == AltinnRepositoryType.Datamodels)
+            {
+                xsdMemoryStream.Position = 0;
+                Json.Schema.JsonSchema jsonSchema = GenerateJsonSchemaFromXsd(xsdMemoryStream);
+                var jsonContent = SerializeJson(jsonSchema);
+
+                await altinnAppGitRepository.WriteTextByRelativePathAsync(
```

Comment:
See if the GetRelativeModelFilePath(string modelName) could be used to avoid duplicating the ChangeExtension logic outside of AltinnAppGitRepository class.

Reply:
fixed